### ISSUE TITLE
SIMD detection - OS support

### ISSUE THREAD

**Issue by amitdo** [tesseract-ocr/tesseract#1348]
The SIMD detection should work on two levels:
1) CPU
2) OS

Currently, we only do (1). We should add (2).

Without this feature, a crash might occur, for example when the CPU was detected as 'supports AVX2', but the OS does not support it.

See also #810 (but please comment here, not there).

**Comment by amitdo** [comment]
https://github.com/tesseract-ocr/tesseract/blob/40f43111e05b3dd2f2f8aeae3aba33016523c881/arch/simddetect.cpp

**Comment by Shreeshrii** [comment]
Related/Duplicate https://github.com/tesseract-ocr/tesseract/issues/519

**Comment by amitdo** [comment]
Hi Shree!

It's a separate issue (not a Duplicate).

**Comment by mirh** [comment]
I believe the only actually OS-sensible x86 extension may be AVX. 
And detection should just be a couple of lines
https://github.com/videolan/x265/blob/3.1.2/source/common/cpu.cpp#L167
https://github.com/FFmpeg/FFmpeg/blob/n4.2/libavutil/x86/cpu.c#L132
https://github.com/PCSX2/pcsx2/blob/v1.5.0-dev/common/src/x86emitter/cpudetect.cpp#L271

**Comment by stweil** [comment]
@mirh, thank you for those hints. Pull request #2620 tries to implement the OS check. Maybe someone has Windows XP running and can test it?

**Comment by mirh** [comment]
@zxcv8553172

**Comment by hglee** [comment]
tesseract 4.1 crashes for old CPU on Windows.

- tesseract 4.1
- Built static by vcpkg with VS2019 on Intel I7 8650U, Windows 10 1903 64-bit
- Machine 1: Intel I5 3570, Windows 7 SP1 64-bit
- Machine 2: Intel I7 2600, Windows 10 1903 64-bit
- Machine 3: Intel Xeon E5 2680, Windows 10 1903 64-bit (under VMWare)
- Machine 4: Intel I7 Q820, Windows 7 SP1 64-bit
- illegal instruction exception(0xc000001d) on event viewer
- Machine 1, 2, and 3 only crashes on OCR running.
- Machine 4 crashes directly even for version command.
- When I remove AVX, AVX2 path(see attached), it works on machine 2, 3. Crashes on machine 4 on version command. Sorry for no debug information.

[fix-simd-remove.txt](https://github.com/tesseract-ocr/tesseract/files/3577146/fix-simd-remove.txt)

- With modified #2620 patch(see attached), still crashes on machine 3, 4. Not tested for machine 1, 2.

[fix-simd-2620.txt](https://github.com/tesseract-ocr/tesseract/files/3577145/fix-simd-2620.txt)

On machine 3, windbg catch this (x86 debug static build)

```
(3750.2534): Illegal instruction - code c000001d (first chance)
(3750.2534): Illegal instruction - code c000001d (!!! second chance !!!)
*** WARNING: Unable to verify checksum for tesseract.exe
eax=00000000 ebx=017e6000 ecx=00000000 edx=00000000 esi=0094ac81 edi=0193ec68
eip=00bb7492 esp=0193ebe8 ebp=0193ebf4 iopl=0         nv up ei pl nz na pe nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00010206
tesseract!tesseract::RecodedCharID::RecodedCharIDHash::operator()+0x52:
00bb7492 c4e271f7d0      shlx    edx,eax,ecx
```

On machine 4, windbg catch this (x86 debug static build)

```
(1a3c.1564): Illegal instruction - code c000001d (first chance)
(1a3c.1564): Illegal instruction - code c000001d (!!! second chance !!!)
*** WARNING: Unable to verify checksum for tesseract.exe
eax=00000004 ebx=756e6547 ecx=00000000 edx=0000000b esi=003efc20 edi=003efc3c
eip=011c65d5 esp=003efbe8 ebp=003efc3c iopl=0         nv up ei pl zr na pe nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00010246
tesseract!tesseract::SIMDDetect::SIMDDetect+0x65:
011c65d5 0f01d0          xgetbv
```

**Comment by amitdo** [comment]
>Machine 4: Intel I7 **Q820**, Windows 7 SP1 64-bi

820QM?
It does not support AVX, seems to support SSE4.1.

The 1-3 CPUs support AVX but not AVX2.

Windows 7 SP1 supportS AVX (not sure about AVX2)

**Comment by hglee** [comment]
> 
> 
> > Machine 4: Intel I7 **Q820**, Windows 7 SP1 64-bi
> 
> Q8200?
> 820QM?
> Both CPUs do not support AVX, seem to support SSE4.1.
> 
> The 1-3 CPUs support AVX but not AVX2.
> 
> Windows 7 SP1 supports AVX (not sure about AVX2)

Windows system info says I7 Q820, maybe 820QM, very old CPU.

**Comment by hglee** [comment]
Sorry for my fault. When I built binary from my machine, I7 8650U, it was built using AVX2 flag by cmake.

So it may need to edit OptimizeForArchitecture.cmake not simddetect.cpp if you want to support multiple machines.

I attached edited patch. After that it works on machine 3, 4 (did not applied #2620 patch).

[fix-optimize.txt](https://github.com/tesseract-ocr/tesseract/files/3577654/fix-optimize.txt)

**Comment by stweil** [comment]
That looks like a regression for CMake builds introduced by commit 21838b0e7a3b4915e88a963b7523e72a452e74bb. @zdenop, do we need that commit? It looks like the resulting executables only run on a CPU which is compatible with the CPU used for building.

**Comment by zdenop** [comment]
That commit should optimize compiler setting for architecture and were originally developed for [SIMD Vector Classes for C++ ](https://github.com/VcDevel/Vc). When I was looking for cmake detection of AVX2, AVX... most of reliable sources pointed to this project.

I expected that tesseract runtime check of presence for AVX instructions will handle situation when support in CPU is missing, but library was build with such support... Maybe I oversaw something, but at the moment I have no time for investigation...

**Comment by mirh** [comment]
> Windows 7 SP1 supports AVX (not sure about AVX2)

AFAIK OS support is only needed when new "kinds" of registers are introduced, not just when you extend already existing ones. 

> I expected that tesseract runtime check of presence for AVX instructions will handle situation when support in CPU is missing, but library was build with such support

If the compiler is explicitly told to optimize binaries for X architecture, of course whatever your code is won't stop it. 
In a perfect world that file/step would be unneeded because they would already have all the various SIMD branches compiled in, and they'd just have to select the right ones at runtime.

**Comment by hglee** [comment]
> 
> 
> @mirh, thank you for those hints. Pull request #2620 tries to implement the OS check. Maybe someone has Windows XP running and can test it?

I tested on Windows XP today, but just crashed on start at xgetbv.

Same binary works on Windows 10 x64.

* Intel I7 2600, Windows XP SP3 32bit
* Above fix-simd-2620, fix-optimize patch applied to 4.1.0 release
* vcpkg ports and script modified for XP (need v141_xp toolset): [vcpkg_tesseract_4.1_xp.txt](https://github.com/tesseract-ocr/tesseract/files/3612617/vcpkg_tesseract_4.1_xp.txt)
* windbg stack trace for x86-static-debug

```
FAULTING_IP: 
tesseract!tesseract::SIMDDetect::SIMDDetect+65 [C:\Workspace\vcpkg\buildtrees\tesseract\src\4.1.0-250eb342b6\src\arch\simddetect.cpp @ 120]
004a5885 0f01d0          xgetbv

EXCEPTION_RECORD:  ffffffff -- (.exr 0xffffffffffffffff)
ExceptionAddress: 004a5885 (tesseract!tesseract::SIMDDetect::SIMDDetect+0x00000065)
   ExceptionCode: c000001d (Illegal instruction)
  ExceptionFlags: 00000000
NumberParameters: 0

FAULTING_THREAD:  00000b40

DEFAULT_BUCKET_ID:  WRONG_SYMBOLS

PROCESS_NAME:  tesseract.exe

ADDITIONAL_DEBUG_TEXT:  
Use '!findthebuild' command to search for the target build information.
If the build information is available, run '!findthebuild -s ; .reload' to set symbol path and load symbols.

FAULTING_MODULE: 7c930000 ntdll

DEBUG_FLR_IMAGE_TIMESTAMP:  5d7cb80c

ERROR_CODE: (NTSTATUS) 0xc000001d - {

EXCEPTION_CODE: (NTSTATUS) 0xc000001d - {

MOD_LIST: <ANALYSIS/>

PRIMARY_PROBLEM_CLASS:  WRONG_SYMBOLS

BUGCHECK_STR:  APPLICATION_FAULT_WRONG_SYMBOLS

LAST_CONTROL_TRANSFER:  from 0045595d to 004a5885

STACK_TEXT:  
0012ff30 0045595d 0012ff4c 00b79179 00455950 tesseract!tesseract::SIMDDetect::SIMDDetect+0x65 [C:\Workspace\vcpkg\buildtrees\tesseract\src\4.1.0-250eb342b6\src\arch\simddetect.cpp @ 120]
0012ff38 00b79179 00455950 00c00540 1bf90e54 tesseract!`dynamic initializer for 'tesseract::SIMDDetect::detector''+0xd [C:\Workspace\vcpkg\buildtrees\tesseract\src\4.1.0-250eb342b6\src\arch\simddetect.cpp @ 56]
0012ff4c 00a8a8ce 00c00000 00c00ae4 1bf90ea8 tesseract!_initterm+0x49 [d:\th\minkernel\crts\ucrt\src\appcrt\startup\initterm.cpp @ 22]
0012ffb0 00a8a7fd 0012ffc0 00a8ab68 0012fff0 tesseract!__scrt_common_main_seh+0xbe [d:\agent\_work\3\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl @ 256]
0012ffb8 00a8ab68 0012fff0 7c817067 01a7f6ee tesseract!__scrt_common_main+0xd [d:\agent\_work\3\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl @ 331]
0012ffc0 7c817067 01a7f6ee 01a7f75c 7ffd8000 tesseract!mainCRTStartup+0x8 [d:\agent\_work\3\s\src\vctools\crt\vcstartup\src\startup\exe_main.cpp @ 17]
WARNING: Stack unwind information not available. Following frames may be wrong.
0012fff0 00000000 0040ac36 00000000 78746341 kernel32!RegisterWaitForInputIdle+0x49


FAILED_INSTRUCTION_ADDRESS: 
tesseract!tesseract::SIMDDetect::SIMDDetect+65 [C:\Workspace\vcpkg\buildtrees\tesseract\src\4.1.0-250eb342b6\src\arch\simddetect.cpp @ 120]
004a5885 0f01d0          xgetbv

FOLLOWUP_IP: 
tesseract!tesseract::SIMDDetect::SIMDDetect+65 [C:\Workspace\vcpkg\buildtrees\tesseract\src\4.1.0-250eb342b6\src\arch\simddetect.cpp @ 120]
004a5885 0f01d0          xgetbv
```

**Comment by mirh** [comment]
You [need to](https://software.intel.com/sites/default/files/managed/9e/bc/64-ia-32-architectures-optimization-manual.pdf#page=191) check OSXSAVE cpuid, before you can use xgetbv.

**Comment by stweil** [comment]
@hglee, @mirh, does it work on XP with [this additional commit](https://github.com/stweil/tesseract/commit/296fa4347ee96ae3e1dd297f31808c4f714095ee)?

**Comment by hglee** [comment]
> 
> 
> @hglee, @mirh, does it work on XP with [this additional commit](https://github.com/stweil/tesseract/commit/296fa4347ee96ae3e1dd297f31808c4f714095ee)?

It did not crashed on xgetbv but automatically select AVX and crashed on DotProductAVX.

When I modified to check xmm_ymm_state_enabled for AVX, it works on XP: [fix-osxsave.txt](https://github.com/tesseract-ocr/tesseract/files/3613493/fix-osxsave.txt)

**Comment by stweil** [comment]
Thank you. Git master should work now with XP. Maybe you want to give it a try and can report the result here.

**Comment by zdenop** [comment]
@hglee: it this fixed? seems like fix-osxsave.txt is not applied (do we need it?)...

fix-optimize.txt is IMO problematic: you removed support for AVX and AVX2 even system would support it (e.g. [AVX2 in VS includes support also for SSE2](https://devblogs.microsoft.com/cppblog/avx2-support-in-visual-studio-c-compiler/)) More reasonable would be that builder could manually select which optimization (/arch:) should be used...

**Comment by mirh** [comment]
It was fixed with https://github.com/tesseract-ocr/tesseract/commit/408d6e8b72c1a3b016fc62556f4639b5445c0e07?
The AVX switches for the AVX opt files weren't touched.

**Comment by hglee** [comment]
> 
> 
> @hglee: it this fixed? seems like fix-osxsave.txt is not applied (do we need it?)...
> 
> fix-optimize.txt is IMO problematic: you removed support for AVX and AVX2 even system would support it (e.g. [AVX2 in VS includes support also for SSE2](https://devblogs.microsoft.com/cppblog/avx2-support-in-visual-studio-c-compiler/)) More reasonable would be that builder could manually select which optimization (/arch:) should be used...

I did not try master branch, but v4.1.0 works with that on XP.

fix-optimize.txt was fix quick to disable compiler flags. That maybe be VC++ compiler problem, because intel compiler can generate optimized code for multi architecture by internal branch. So if you need to distribute binary for most common machine with VC++ compiler, you need to disable some flags.

**Comment by mirh** [comment]
Gcc (and llvm?) also has function multiversioning, but perhaps I'm digressing. 

TL;DR is there someone reporting any problem whatsoever now?
